BJS2-16323

fokinms opened the pull request:
createGoal();
updateGoal();
deleteGoal();
findSubtaskByGoalId();
findGoalsByUserId();

create Mapper, Filter with filters classes;
create GoalDTO;
create GoalFilterDTO;

fraifree committed:
create SubscriptionService
create SubscriptionController
with methods:
followUser();
unFollowUser();
getFollowers();
getFollowing();
getFollowersCount();
getFollowingCount();
create UserFilter interface with filters;
refactor SubscriptionRepository;
add SubscriptionControllerTest and SubscriptionServiceTest

fokinms committed:
create GoalService, GoalController with methods:
createGoal();
updateGoal();
deleteGoal();
findSubtaskByGoalId();
findGoalsByUserId();

create Mapper, Filter with filters classes;
create GoalDTO;
create GoalFilterDTO;

fraifree commented on src/main/java/school/faang/user_service/service/goal/GoalDescriptionFilter.java:
```diff
@@ -0,0 +1,22 @@
+import school.faang.user_service.entity.goal.Goal;
+
+import java.util.stream.Stream;
+
```
я бы вынес фильтры в отдельный пакет, на мой взгляд в сервисах должны лежать именно сервисы

fokinms replied:
Done

fraifree commented on src/main/java/school/faang/user_service/service/goal/GoalService.java:
```diff
@@ -0,0 +1,95 @@
+
+@Service
+@RequiredArgsConstructor
+public class GoalService {
```
В методах сервиса есть похожая логика по валидации, я бы подумал как это вывести в отдельный приватный метод

fokinms replied:
Done

Commit:
fixed gradle file

fraifree committed:
fix after review

fokinms committed:
Merge branch 'refs/heads/griffon-master-bc5' into fms-BJS2-16323

fokinms committed:
create GoalService, GoalController with methods:
createGoal();
updateGoal();
deleteGoal();
findSubtaskByGoalId();
findGoalsByUserId();

create Mapper, Filter with filters classes;
create GoalDTO;
create GoalFilterDTO;

create GoalControllerValidator;
create GoalServiceValidator;

create custom exception class;

changed UserRepository & GoalRepository extending;

fokinms committed:
Add Swagger:
Add dependencies in build.gradle.kts;
create OpenApiConfig

fokinms committed:
Add Swagger annotations for GoalDto & GoalController

InnMish commented on src/main/java/school/faang/user_service/controller/goal/GoalController.java:
```diff
@@ -0,0 +1,71 @@
+@Tag(name = "GoalController", description = "Goal API")
+@RestController
+@RequiredArgsConstructor
+@RequestMapping(value = "/goals")
```
тут value можно опустить, сразу значение в кавычках пиши) 
мелочь, но зачем лишнее слово, если можно без него)

fokinms replied:
Done

InnMish commented on src/main/java/school/faang/user_service/controller/goal/GoalController.java:
```diff
@@ -0,0 +1,71 @@
+    @PostMapping(value = "{userId}/goals/create")
+    @ResponseStatus(HttpStatus.CREATED)
+    public GoalDto createGoal(@Valid  @Parameter(description = "user id") @PathVariable Long userId, @RequestBody GoalDto goalDto) {
+        goalControllerValidator.validateCreation(goalDto);
```
так-так-тааак
давай не будем столько мелких классов лепить 
давай сделаем один общий валидатор GoalValidator и там разные методы просто 
иначе очень много мелких пакетов и классов слишком узких 

одного вполне достаточно для целей валидации. Нам же главное просто скрыть какую-то реализацию из общего класса, чтобы тут не было деталей всяких

fokinms replied:
Done

InnMish commented on src/main/java/school/faang/user_service/controller/goal/GoalController.java:
```diff
@@ -0,0 +1,71 @@
+    @Operation(summary = "Find subtasks", description = "find subtasks by goal id")
+    @GetMapping(value = "/goals/{goalId}/subtasks")
+    @ResponseStatus(HttpStatus.OK)
+    private List<GoalDto> findSubtaskByGoalId(@Valid  @Parameter(description = "goal id") @PathVariable Long goalId, @RequestBody GoalFilterDto filters) {
```
вот такие длинные строчки-гусеницы дели на две строчки, иначе это сложно воспринимается 
в идее есть разграничение вертикальное, полосочка такая
вот за нее старайся не выходить, переноси на след.строчку

fokinms replied:
Done

InnMish commented on src/main/java/school/faang/user_service/controller/goal/validation/GoalControllerValidator.java:
```diff
@@ -0,0 +1,35 @@
+        validateDescription(goalDto);
+    }
+
+    public void validateUpdating(GoalDto goalDto) {
```
так подождите) 
а почему тут два одинаковых метода?) 

Ты же можешь назвать его, например, validateGoal и дергать его два раза) 

Никогда не пиши одинаковый код, переиспользуй всегда то, что можешь переиспользовать

fokinms replied:
Done

InnMish commented on src/main/java/school/faang/user_service/controller/goal/validation/GoalControllerValidator.java:
```diff
@@ -0,0 +1,35 @@
+    }
+
+    private void validateTitle(GoalDto goalDto) {
+        if (goalDto.getTitle().isEmpty()) {
```
ты же аннотацию Valid используешь, давай вот эту проверку тайтла сделаем на уровне дто аннотацией

fokinms replied:
Done

InnMish commented on src/main/java/school/faang/user_service/controller/goal/validation/GoalControllerValidator.java:
```diff
@@ -0,0 +1,35 @@
+    private void validateTitle(GoalDto goalDto) {
+        if (goalDto.getTitle().isEmpty()) {
+            log.info("title is empty");
+            throw new InvalidRequestParams("Title is empty");
```
Сделай общую ошибку DataValidationException и на все ошибки валидации ее кидай 
Потому что такая ошибка текущая довольно узкая, только про недопустимые параметры 
А у нас валидации в коде разные могут быть, и чтобы ошибка везде подходила, давай сделаем ее более общей

InnMish commented on src/main/java/school/faang/user_service/controller/goal/validation/GoalControllerValidator.java:
```diff
@@ -0,0 +1,35 @@
+    }
+
+    private void validateDescription(GoalDto goalDto) {
+        if (goalDto.getDescription().isEmpty()) {
```
тут тоже давай на уровне дто проверять аннотацией

InnMish commented on src/main/java/school/faang/user_service/entity/goal/GoalStatus.java:
```diff
@@ -1,6 +1,7 @@
 
-public enum GoalStatus {
+public enum
+GoalStatus {
```
а зачем тут перенос на другую строчку?

InnMish replied:
не исправлено

fokinms replied:
Done

InnMish commented on src/main/java/school/faang/user_service/mapper/GoalDtoMapper.java:
```diff
@@ -0,0 +1,44 @@
+
+    List<GoalDto> toDtos(List<Goal> goals);
+
+    @Mapping(target = "parent", ignore = true)
+    @Mapping(target = "invitations", ignore = true)
+    @Mapping(target = "users", ignore = true)
+    @Mapping(target = "skillsToAchieve", ignore = true)
```
если все игноришь, unmappedTargetPolicy = ReportingPolicy.IGNORE) вот это же их и так игнорит 
еще такой же параметр есть на sourcePolicy

fokinms replied:
Done

InnMish commented on src/main/java/school/faang/user_service/service/goal/GoalService.java:
```diff
@@ -0,0 +1,100 @@
+
+    @Transactional
+    public GoalDto createGoal(long userId, GoalDto goalDto) {
+        goalServiceValidator.validateCreation(userId, goalDto);
```
соответственно это у тебя тоже будет в общем валидатор-классе

InnMish commented on src/main/java/school/faang/user_service/service/goal/GoalService.java:
```diff
@@ -0,0 +1,100 @@
+        goalServiceValidator.validateCreation(userId, goalDto);
+        Goal goal = goalDtoMapper.toEntity(goalDto);
+        goal.setStatus(GoalStatus.ACTIVE);
+        goal.setUsers(userRepository.findAllById(goalDto.getUsersId()));
```
в идеале дергать методы репозитория через сервис 
то есть если дергаешь метод userRepository, то результат собирай в userService, а сюда инжекти сервис

InnMish commented on src/main/java/school/faang/user_service/service/goal/GoalService.java:
```diff
@@ -0,0 +1,100 @@
+        Goal goal = goalDtoMapper.toEntity(goalDto);
+        goal.setStatus(GoalStatus.ACTIVE);
+        goal.setUsers(userRepository.findAllById(goalDto.getUsersId()));
+        goal.setSkillsToAchieve(skillRepository.findAllById(goalDto.getSkillsToAchieveId()));
```
а это через SkillService

fokinms replied:
Done

InnMish commented on src/main/java/school/faang/user_service/service/goal/GoalService.java:
```diff
@@ -0,0 +1,100 @@
+    @Transactional
+    public GoalDto updateGoal(long goalId, GoalDto goalDto) {
+        goalServiceValidator.validateUpdating(goalId, goalDto);
+        Goal goal = goalDtoMapper.toEntity(goalDto);
```
Хе, нет 
чтобы обновить то, что уже есть в бд, это из бд надо достать и засетить туда измененные поля

fokinms replied:
Апдейт с помощью метода save интерфейса Repository проверяет наличие id cущности в базе, и принимает решение сделать merge или persist. в случае наличия сущности он автоматически подтянет все поля, которые отличаются по значению и сохранит их. Возможно, что то неправильно понял. Можно подробнее как правильно сделать и почему такой способ не подойдет?

InnMish replied:
Все верно пишешь 
Если так и было - ок, невнимательно посмотрела может

InnMish commented on src/main/java/school/faang/user_service/service/goal/validation/GoalServiceValidator.java:
```diff
@@ -0,0 +1,81 @@
+    private final GoalRepository goalRepository;
+
+
+    public void validateCreation(long userId, GoalDto goalDto) {
+        validateUserExistence(userId);
+        validateActiveGoalsCount(userId);
+        validateSkillsExistence(goalDto);
+    }
+
+    public void validateUpdating(long goalId, GoalDto goalDto) {
+        validateGoalExistence(goalId);
+        validateStatus(goalId);
+        validateSkillsExistence(goalDto);
+    }
```
не дублируем методы

fokinms replied:
несколько мелких методов собирают логику валидации для конкретной задачи. их всячески можно комбинировать, комбинации разные в этих двух задачах. как их стоит оформить? не уловил

InnMish commented on src/main/java/school/faang/user_service/service/goal/validation/GoalServiceValidator.java:
```diff
@@ -0,0 +1,81 @@
+        validateSkillsExistence(goalDto);
+    }
+
+    public void validateDeleting(long goalId) {
+        validateGoalExistence(goalId);
+    }
+
+    public void validateFindingSubtasksByGoalId(long goalId) {
+        validateGoalExistence(goalId);
+    }
+
+    public void validateFindingGoalsByUserId(long userId) {
+        validateUserExistence(userId);
+    }
```
а зачем тебе такие методы 
ты же напрямую можешь дергать те, что внутри 
а то обертка над методами получилась бесполезная

fokinms replied:
Done

InnMish commented on src/main/java/school/faang/user_service/service/goal/validation/GoalServiceValidator.java:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    private void validateActiveGoalsCount(long userId) {
+        if (goalRepository.countActiveGoalsPerUser(userId) > 2) {
```
2 выносим в константу, должно быть понятно, что это за число по названию поля

fokinms replied:
Done

fraifree committed:
fix after review
create UserService and UserController for testing request by Post Service module
create test by filters

fokinms committed:
Corrected comments
Create application.properties file for constants
Create SkillService & UserService with necessary methods

fokinms committed:
Correct GoalStatus

InnMish commented on src/main/java/school/faang/user_service/service/goal/GoalService.java:
```diff
@@ -0,0 +1,100 @@
+
+    @Transactional
+    public void deleteGoal(long goalId) {
+        goalValidator.validateGoalExistence(goalId);
```
метод delete ничего не удалит и даже не кинет ошибку, если такой сущности в бд нет 
это норм, можно вообще тут не проверять дополнительно ничего

fokinms replied:
Done

InnMish commented on src/main/resources/application.properties:
```diff
@@ -0,0 +1 @@
+goal_validator.max_active_goals_per_user_for_creating_new_goal = 2
```
почему тут? у тебя же есть ямл-файл cо всеми конфигами, пиши в нем

fokinms replied:
Done

fraifree committed:
update packages

fokinms committed:
Merge branch 'refs/heads/griffon-master-bc5' into fms-BJS2-16323

fokinms committed:
Fix comments

InnMish commented on src/main/java/school/faang/user_service/controller/GoalController.java:
```diff
@@ -0,0 +1,71 @@
+    private final GoalService goalService;
+
+    @Operation(summary = "Create Goal", description = "create new goal")
+    @PostMapping("{userId}/goals/create")
```
https://discord.com/channels/1102662981743030334/1102662982447669361/1265711557434675231

fokinms replied:
Угумс, согласен. Пользовался здесь файликом FEATURES.md, который лежит в проекте user-service. Там были такие пути указаны. Исправлю, сделаю как надо. Тот файлик, возможно, стоит исправить. Он полезен когда первый раз начинаешь работу с проектом, но в нем море ошибок)

InnMish commented on src/main/java/school/faang/user_service/exception/InvalidRequestParams.java:
```diff
@@ -0,0 +1,8 @@
+
+public class InvalidRequestParams extends RuntimeException {
+
+    public InvalidRequestParams(String message) {
```
Писала вроде бы 
Мы в других задачах в user service используем исключение DataValidationException 
Давай тоже так назовем, чтобы исключение было не только для параметров, а вообще для всех ошибок валидации

fokinms replied:
Свое удалил, подтянул из основной ветки, заюзал

InnMish commented on src/main/java/school/faang/user_service/mapper/goal/GoalDtoMapper.java:
```diff
@@ -0,0 +1,40 @@
+import java.util.List;
+
+@Mapper(componentModel = "spring", unmappedTargetPolicy = ReportingPolicy.IGNORE)
+public interface GoalDtoMapper {
```
GoalMapper просто 
Мы же тут дто в энтити и энтити в дто маппим

fokinms replied:
Логишно. Done

InnMish commented on src/main/resources/application.yaml:
```diff
@@ -51,3 +51,6 @@ services:
     isMocked: true
+
+goal:
+  max_active_goals = 2
```
max_active_goals: 2

yaml стандарт

fokinms replied:
Done

fokinms committed:
Merge branch 'refs/heads/griffon-master-bc5' into fms-BJS2-16323

fokinms committed:
Fix comments
Fix filter bugs
Fix UserService
Fix GoalService bugs
Create GoalServiceTest

InnMish commented on src/main/java/school/faang/user_service/service/GoalService.java:
```diff
@@ -0,0 +1,107 @@
+        if (goal.getStatus() != null && goal.getStatus().equals(GoalStatus.COMPLETED)) {
+            updateUsersAndSkillsWhenGoalCompleted(goal);
+        }
+        if (goal.getUsers() != null) {
```
а это возможно вообще? Проверь, как у тебя маппер работает. Там будет пусто список, а не null

fokinms replied:
не у того объекта проверил, исправил

InnMish commented on src/main/java/school/faang/user_service/service/GoalService.java:
```diff
@@ -0,0 +1,107 @@
+        if (goal.getStatus() != null && goal.getStatus().equals(GoalStatus.COMPLETED)) {
+            updateUsersAndSkillsWhenGoalCompleted(goal);
+        }
+        if (goal.getUsers() != null) {
+            goal.setUsers(userService.findAllById(goalDto.getUserIds()));
+        }
+        goal.setSkillsToAchieve(skillService.findAllById(goalDto.getSkillsToAchieveIds()));
+        return goalMapper.toDto(goalRepository.save(goal));
```
у тебя в методе выше есть точно такой же код 
Можем избавиться от дублирования?

fokinms replied:
Done

InnMish commented on src/test/java/school/faang/user_service/service/GoalServiceTest.java:
```diff
@@ -0,0 +1,238 @@
+                .username("test_user_2")
+                .skills(skills)
+                .build();
+        skill_1 = Skill.builder()
```
поля с нижним подчеркиванием в java не называем

InnMish commented on src/test/java/school/faang/user_service/service/GoalServiceTest.java:
```diff
@@ -0,0 +1,238 @@
+    }
+
+    @Test
+    void createGoal_whenInvalidParameters_thenThrowsException() {
```
testGoalWithWrongParametersThrowsException

InnMish replied:
@displayTest

InnMish commented on src/test/java/school/faang/user_service/service/GoalServiceTest.java:
```diff
@@ -0,0 +1,238 @@
+
+    @Test
+    void createGoal_whenInvalidParameters_thenThrowsException() {
+        doThrow(DataValidationException.class).when(goalValidator).validateCreation(notExistUserId, goalWrongDto);
```
а почему так делаешь? 
Тебе тут надо в createGoal передать notExistUserId, и валидация уже свалится. Тогда ты можешь проверить и тип исключения, и текст его 

Также с дальнейшими валидациями:
countActiveGoalsPerUser мокаешь на больше, чем нужно - ловишь ошибку. 

А иначе вот ясности в этом тесте не особо много 
Просто стаббим, говорим, что выпадет ошибка, а потом убеждаемся в том, что она реально падает) 
Давай лучше по отдельности участки все проверять. 

Я бы рекомендовала в сервисе проверять всю основную логику с валидными данными, а на валидатор написать отдельный класс тестов

fokinms replied:
Валидатор замокан, он не свалится сам. Так делаю, т.к. валидатор в проверках возвращает void, поведение задаю самому валидатору, а проверяю сработало ли вызывая тестируемый метод. Если есть способы как то по другому сделать, я бы их глянул. В данном случае не нашел других вариантов

InnMish commented on src/test/java/school/faang/user_service/service/GoalServiceTest.java:
```diff
@@ -0,0 +1,238 @@
+        when(goalMapper.toEntity(goalDto)).thenReturn(goal);
+        when(userService.findAllById(goalDto.getUserIds())).thenReturn(users);
+        goalService.updateGoal(existGoalId, goalDto);
+        Assertions.assertEquals(users, goal.getUsers());
```
тут бы еще проверять, что save вернул ожидаемый результат

fokinms replied:
Надо ли? Я здесь тестирую сервис, не репо. Специально не стал пробивать все строки подряд, старался трогать только то, что непосредственно к сервису относится

fokinms replied:
сделал дополнительную проверку, что маппер запускается для возврата из метода новой ДТО, после сохранения преобразованной сущности в репозиторий

InnMish commented on src/test/java/school/faang/user_service/service/GoalServiceTest.java:
```diff
@@ -0,0 +1,238 @@
+    void updateGoal_whenValidParameters_thenSaveGoal() {
+        when(goalMapper.toEntity(goalDto)).thenReturn(goal);
+        goalService.updateGoal(existGoalId, goalDto);
+        verify(goalRepository).save(goal);
```
проверить, что save вернул то, что ожидаем 
маппер рекомендую не мокать, а Spy на него вешать и проверять, что отрабатывает, как ожидаем 
Иначе тут вся логика маппера непокрыта осталась

fokinms replied:
Done

InnMish commented on src/test/java/school/faang/user_service/service/GoalServiceTest.java:
```diff
@@ -0,0 +1,238 @@
+        goalService.findGoalsByUserId(existUserId, filter);
+        verify(goalMapper).toDtos(any());
+    }
+}
```
отдельно еще надо тесты на фильтры написать 
Посмотри в мастере уже есть тесты от Максима, можно взять за пример)

fokinms replied:
Гляну. Я хотел с тобой минут 10 по тестам в ПП пообщаться. Нужна концепция на постоянку)

fokinms committed:
Fix comments
Fix mapper
Create Filters tests and combined filter test
Create GoalValidatorTest
